### PR TITLE
fix(ai-chat): close conversation-identity race causing sends to land in the wrong conversation

### PR DESCRIPTION
**Pull request by 2witstudios**
## Summary

Sending a message to an existing conversation could create a new one instead of appending, and starting a new conversation could append to the previous one instead. Root cause: conversation identity was tracked independently in `useConversations.ts`, `usePageAgentDashboardStore.ts`, `GlobalChatContext.tsx`, and `AiChatView.tsx` local state, each writing it asynchronously after a network round trip — a send fired mid-transition could carry a stale id, and a stale async response could overwrite an identity the user had already moved past.

- **Pure state machine** (`conversation-identity.ts`): `idle → resolving → ready/error`. `IDENTITY_SET` always wins regardless of current state; a stale `RESOLVED`/`RESOLVE_FAILED` that arrives after a newer identity was set is a guaranteed no-op. One shared, independently-tested primitive.
- **Client-generated ids**: conversation creation generates its id synchronously (cuid2) instead of waiting on the server to return one — the create/persist POST becomes fire-and-forget and idempotent. The page-agents create route now accepts (and the merged `pu/cli-login` ownership-guard fix protects) a client-supplied id.
- **Every surface wired to the same reducer via the mechanism appropriate to its container** — not one-size-fits-all:
  - `AiChatView.tsx`: the new shared `useConversationIdentity` hook (single-instance component).
  - `usePageAgentDashboardStore.ts`: the reducer called directly inside Zustand actions (a store isn't a React component, so the hook doesn't apply — the plain reducer function does).
  - `GlobalChatContext.tsx`: a local `useReducer` in the provider (shared by every consumer via context, not duplicated per-consumer).
  - `usePageAgentSidebarState.ts`: a 4th independent surface (sidebar agent chat) discovered during proactive review — same reducer, same Zustand-action wiring pattern as the dashboard store.
- **`AiChatView` keyed by `page.id`** in `CenterPanel.tsx`, matching the existing pattern already used for `DocumentView`/`SheetView`/etc. — the component can receive a new `page` prop without unmounting, and React's own documented pattern for resetting state on a new subject is `key`, not a hand-rolled reset-on-prop-change effect.
- **Folds in a second root cause**: `AiChatView`'s init effect misclassified a transient list-fetch failure as "no conversations exist," silently replacing a real conversation with a fresh disconnected one. It now surfaces a retry banner instead.

### Hardening pass (proactive + reviewer-driven, post-initial-implementation)

- Fixed a double-fetch on `AiChatView` history-select (skip-token set after, not before, adopting identity) and restored history-list memoization via a stable `onSelectConversation`.
- `resolveConversation`'s messages-prefetch exception no longer misclassifies as a full identity-resolution failure when the conversation itself was already found.
- Guarded `loadMostRecentConversation`'s catch/no-conversation-found paths, the late-joiner stream-reconciliation `setIdentity` call, and a stale Zustand no-op `set()` — all against clobbering a newer identity the user has since moved to.
- Added a decoupled `isConversationMessagesLoading` / `isMessagesLoading` flag (store, context, sidebar-state hook, both consumer views) — adopting identity synchronously closed the race but also removed the window that used to cover the messages fetch, so a conversation switch could otherwise flash the previous conversation's messages under the new identity with no loading indicator.
- Client-supplied `conversationId` on the create route is now validated with `isCuid()` before being trusted as a row id.
- `GlobalChatContext`'s non-ok messages-fetch path now clears `initialMessages` (previously only the thrown-exception path did).

## Test plan

- [x] `bun run typecheck` — clean
- [x] 227+ new/updated unit tests across every touched file, all green (reducer, hook, store, context, sidebar-state hook, `AiChatView`, route)
- [x] `bun run --filter web test` — 736/740 files green project-wide (4 pre-existing failures unrelated to this change: a missing Postgres test role and a midnight-boundary date test)
- [ ] Manual smoke test in a running instance: New Chat → immediate send lands in the new conversation; history-select → immediate send lands in the selected conversation; simulated network failure on load surfaces a retry banner instead of silently starting a disconnected conversation

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01JT8umuwBmmtKiwmBhyq5uH

## Summary by CodeRabbit

* **New Features**
  * Conversation threads now appear immediately when created, with a smoother transition between chats.
  * Chat views show a clearer loading state while messages are being fetched.

* **Bug Fixes**
  * Fixed cases where switching conversations could briefly show the wrong messages.
  * Improved reliability when loading, retrying, or restoring chat conversations.
  * Added better error handling so chat areas can recover more gracefully when conversation data fails to load.

### CODE REVIEW

File: apps/web/src/components/layout/middle-content/page-views/ai-page/AiChatView.tsx [hunk 1]
```diff
@@ -1077,7 +1111,13 @@ const AiChatView: React.FC<AiChatViewProps> = ({ page }) => {
           <PageAgentHistoryTab
             conversations={conversations}
             currentConversationId={currentConversationId}
-            onSelectConversation={loadConversation}
+            onSelectConversation={(conversationId) => {
+              // Adopt the selected id immediately — before the messages fetch even
+              // starts — so a send fired right after clicking a history item can't
+              // race and land under the previous conversation.
+              setIdentity(conversationId);
+              void loadConversation(conversationId);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/middle-content/page-views/ai-page/AiChatView.tsx]: Fixed in 564c0decc: `handleSelectConversation` now sets `skipLoadEffectRef.current = conversationId` *before* calling `setIdentity`, so the load-on-select effect sees the skip token already in place when it reacts to the identity change and no longer fires its own redundant fetch. `onSelectConversation` was also hoisted into a stable `useCallback` (was previously a fresh inline arrow every render, which separately broke the history list's memoization).

Verified with a regression test (`AiChatView.test.tsx`, "given a conversation is selected from history and onConversationLoad supplies messages, should NOT double-fetch that conversation over the network") that fails without the fix and passes with it.

Leaving this thread open for verification rather than resolving it myself, since the fix landed after this review comment was posted.

File: apps/web/src/app/api/ai/page-agents/[agentId]/conversations/route.ts [hunk 2]
```diff
@@ -178,12 +178,17 @@ export async function POST(
...
 
-    // Generate new conversation ID using createId
-    const conversationId = createId();
+    // Prefer a client-generated id (cuid2) so the caller knows its conversation
+    // id synchronously, before this request resolves. Fall back to generating
+    // one server-side for callers that don't supply it.
+    const conversationId: string =
+      typeof body.conversationId === 'string' && body.conversationId.length > 0
+        ? body.conversationId
+        : createId();
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/page-agents/[agentId]/conversations/route.ts:192]: Fixed in f8dc36197: now validates `body.conversationId` with `isCuid()` before trusting it, falling back to `createId()` on a malformed value. Added tests for both the valid-CUID2 and malformed-id cases.

File: apps/web/src/contexts/GlobalChatContext.tsx [hunk 3]
```diff
@@ -95,27 +131,32 @@ export function GlobalChatProvider({ children }: { children: ReactNode }) {
...
         const loadedMessages = Array.isArray(messageData) ? messageData : messageData.messages || [];
         setInitialMessages(loadedMessages);
-        setCurrentConversationId(conversationId);
-        conversationState.setActiveConversationId(conversationId);
-        setIsInitialized(true);
       } else {
         console.error('Failed to load conversation:', conversationId);
-        setIsInitialized(true);
       }
+      setIsMessagesLoading(false);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/contexts/GlobalChatContext.tsx:155]: Fixed in f8dc36197: the non-ok branch now calls `setInitialMessages([])`, matching the existing catch branch. Added a regression test asserting the previous conversation's messages don't remain visible after a non-ok response.

File: apps/web/src/stores/page-agents/usePageAgentDashboardStore.ts [hunk 4]
```diff
@@ -181,63 +235,66 @@ export const usePageAgentDashboardStore = create<AgentState>()((set, get) => ({
...
+      await createAgentConversation(agent.id, conversationId);
     } catch (error) {
       console.error('Failed to create new conversation:', error);
       toast.error('Failed to create new conversation');
-      set({ isConversationLoading: false });
-      return null;
     }
+
+    return conversationId;
   },
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/stores/page-agents/usePageAgentDashboardStore.ts:299]: Fixed in f8dc36197: `createNewConversation`'s `set(...)` block now also resets `isConversationMessagesLoading: false`. Added a regression test covering the scenario you described — a `loadConversation` fetch still in-flight when `createNewConversation` fires, confirming the flag clears immediately and stays cleared when the stale fetch resolves afterward.

File: apps/web/src/components/layout/middle-content/page-views/ai-page/AiChatView.tsx [hunk 5]
```diff
@@ -132,12 +134,177 @@ const AiChatView: React.FC<AiChatViewProps> = ({ page }) => {
   // Always reflects the current page.id so async callbacks can detect stale pages
   const pageIdRef = useRef(page.id);
   useEffect(() => { pageIdRef.current = page.id; }, [page.id]);
+  // Always reflects the current identity's conversationId (kept in sync by an
+  // effect further below, once currentConversationId is derived) so async
+  // reconciliation callbacks (e.g. the late-joiner sync) can detect that the
+  // user has since switched away before applying a stale setIdentity call.
+  const currentConversationIdRef = useRef<string | null>(null);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/middle-content/page-views/ai-page/AiChatView.tsx]: Evaluated this and decided not to apply it. Moved the update into a `useEffect` (matching `pageIdRef`) and re-ran the suite — it broke 2 existing tests ("given fireComplete fires with stream.conversationId matching the persisted conversation while currentConversationId is the page-scoped default..." and the sibling "...should refresh the conversation list..." test), because those tests trigger the late-joiner callback synchronously without an intervening `act()`-wrapped flush, and the ref-sync effect hadn't committed yet by the time the guard read it.

More fundamentally: this component doesn't use any concurrent-rendering features (no Suspense, no `startTransition`) that could discard an in-progress render before commit — the hazard a `useEffect`-based ref sync guards against. Meanwhile moving to an effect introduces a real, demonstrated timing gap: an async callback that fires before React flushes a just-scheduled effect reads a stale ref value. So the render-body write is actually the more correct choice here, not just simpler. Left a comment in the code explaining the tradeoff (apps/web/src/components/layout/middle-content/page-views/ai-page/AiChatView.tsx around line 304).
